### PR TITLE
Generating API Docs for both Drone and Server APIs

### PR DESCRIPTION
**Pull request by chrizandr**
Added API Documentation for Server and Drone APIs.

Created generator functions for creating the API Doc, it should be easy to modify and generate the docs for both Server and Drone.

The Server has a copy of the Drone Class, having same properties and operations. It would be better if these are mapped to the properties and operations on the Drone API.
Server also has LogEntry for all the log entries and Area for the area of interest.

Please review

### CODE REVIEW

File: api_docs/server_doc_gen.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                        "API Doc for the server side API",
+                        "API Documentation for the server side system",
+                        "/"+API+"/")
+
+    # Drone operations on server must be linked to same operations on the specific drone
+    drone = drone_doc(API)["supportedClass"][0]
+
+    log = hydra_class("LogEntry", "LogEntry", "Class for a log entry", API)
+    # Subject
+    log["supportedProperty"].append(hydra_prop("http://hydrus.com/serverapi/vocab/Drone", "Drone", "true", "true", "true"))
```
Comment (Mec-iS) [inline api_docs/server_doc_gen.py]: we should have a proper getter/setter to append supported properties. Something like: `Log.add_supported_prop()`

File: api_docs/server_doc.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    "possibleStatus": [],
+    "supportedClass": [
+        {
+            "@context": "http://hydrus.com/context/Drone.jsonld",
+            "@id": "http://hydrus.com/serverapi/vocab/Drone",
+            "@type": "Class",
+            "description": "Class for a drone",
+            "supportedOperation": [
+                {
+                    "@id": "/serverapi/change_speed",
```
Comment (Mec-iS) [inline api_docs/server_doc.py]: The id should be `/droneapi/...` I suppose

File: api_docs/server_doc.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            "@context": "http://hydrus.com/context/LogEntry.jsonld",
+            "@id": "http://hydrus.com/serverapi/vocab/LogEntry",
+            "@type": "Class",
+            "description": "Class for a log entry",
+            "supportedOperation": [
+                {
+                    "@id": "/serverapi/log_entry",
+                    "@type": "Operation",
+                    "expects": "http://hydrus.com/LogEntry",
+                    "method": "POST",
```
Comment (xadahiya) [inline api_docs/server_doc.py]: I don't think we need post requests for log_entry, log entries are automatically generated from various drone datastream and status submissions. I think it should be get.

File: api_docs/server_doc.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+                    "required": "true",
+                    "title": "BottomRight",
+                    "writeable": "true"
+                }
+            ],
+            "title": "Area"
+        }
+    ],
+    "title": "API Doc for the server side API"
+}
```
Comment (xadahiya) [inline api_docs/server_doc.py:220]: We'll need to add more endpoints from the design doc, rest everything is fine! 👍

File: api_docs/drone_doc.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                    "method": "PUT",
+                    "possibleStatus": [
+                        {
+                            "code": 200,
+                            "description": "Status Changed"
+                        }
+                    ],
+                    "returns": "null",
+                    "title": "ChangeStatus"
+                },
```
Comment (xadahiya) [inline api_docs/drone_doc.py]: I don't think we can support all `change_speed`, `change_position`, `change_status` PUT operations at a single endpoint. We can create different methods for a single type like for speed we can have `change_speed`, `get_speed` etc but I'm not sure how we can use a single endpoint `drone` to handle different things. 
Here's an example of event class from Markus Lanthler's event api 
```json
{
      "@id": "http://schema.org/Event",
      "@type": "hydra:Class",
      "hydra:title": "Event",
      "hydra:description": null,
      "supportedOperation": [
        {
          "@id": "_:event_replace",
          "@type": "http://schema.org/UpdateAction",
          "method": "PUT",
          "label": "Replaces an existing Event entity",
          "description": null,
          "expects": "http://schema.org/Event",
          "returns": "http://schema.org/Event",
          "statusCodes": [
            {
              "code": 404,
              "description": "If the Event entity wasn't found."
            }
          ]
        },
        {
          "@id": "_:event_delete",
          "@type": "http://schema.org/DeleteAction",
          "method": "DELETE",
          "label": "Deletes a Event entity",
          "description": null,
          "expects": null,
          "returns": "http://www.w3.org/2002/07/owl#Nothing",
          "statusCodes": [
          ]
        },
        {
          "@id": "_:event_retrieve",
          "@type": "hydra:Operation",
          "method": "GET",
          "label": "Retrieves a Event entity",
          "description": null,
          "expects": null,
          "returns": "http://schema.org/Event",
          "statusCodes": [
          ]
        }
      ],
      "supportedProperty": [
        {
          "property": "http://schema.org/name",
          "hydra:title": "name",
          "hydra:description": "The event's name",
          "required": true,
          "readonly": false,
          "writeonly": false
        },
        {
          "property": "http://schema.org/description",
          "hydra:title": "description",
          "hydra:description": "Description of the event",
          "required": true,
          "readonly": false,
          "writeonly": false
        },
        {
          "property": "http://schema.org/startDate",
          "hydra:title": "start_date",
          "hydra:description": "The start date and time of the event in ISO 8601 date format",
          "required": true,
          "readonly": false,
          "writeonly": false
        },
        {
          "property": "http://schema.org/endDate",
          "hydra:title": "end_date",
          "hydra:description": "The end date and time of the event in ISO 8601 date format",
          "required": true,
          "readonly": false,
          "writeonly": false
        }
      ]
    }
```

@Mec-iS ^^
One solution is to create a endpoint that receives orders as objects (see [doc](https://docs.google.com/document/d/1BeLoBjT43qhxcUub0MDQ_7OL6yXPVnVWE7-E-CCC-WA/edit))and further processing can be done internally. For example - 
`/droneapi/receive_order` [POST]
```json
{
    "order_id": 123,
    "speed": 30,
    "destination": "xyz"
}

File: api_docs/drone_doc.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                        {
+                            "code": 200,
+                            "description": "Temperature returned"
+                        }
+                    ],
+                    "returns": "http://schema.org/Float",
+                    "title": "GetTemperature"
+                },
+                {
+                    "@id": ":recharge",
```
Comment (xadahiya) [inline api_docs/drone_doc.py]: Not sure about this one 😅 ,as recharge will be handled internally by the drone who will make POST requests?

**Review by xadahiya — COMMENTED** [review]
Nice work Chris, however I think we should discuss a few points. See my comments below. Cheers 😄

File: api_docs/server_doc.py (hunk 3 above)
Comment (chrizandr) [reply api_docs/server_doc.py]: Okay, I'll remove that.

File: api_docs/drone_doc.py (hunk 6 above)
Comment (chrizandr) [reply api_docs/drone_doc.py]: I think what we discussed was that the server will check the battery status and see how far a drone can go before it runs out. Based on that the server must calculate the nearest chargin station, see how much battery is needed to be able to reach there and decide on a path for the drone to take. All this computation on the drone will be a little too much, better to let the server handle this and send the instructions to the drone.

File: api_docs/drone_doc.py (hunk 5 above)
Comment (chrizandr) [reply api_docs/drone_doc.py]: I'll modify everything into a status object, and we can send a new status object every time we need to modify some feild or the other. Something similar to receive_order I think

**Pull request merged**